[compiler-rt] Propagate sysroot from CMake to msan tests

zeroomega opened the pull request:
Some msan tests requires rpc/xdr.h, which is no longer available in some newer glibc. In build systems with hermetic sysroot set through CMake, this issue can be solved by setting the sysroot to CMAKE_SYSROOT for these tests. This patch implements it.

zeroomega committed:
[compiler-rt] Propagate sysroot from CMake to msan tests

Some msan tests requires rpc/xdr.h, which is no longer available
in some newer glibc. In build systems with hermetic sysroot set through
CMake, this issue can be solved by setting the sysroot to CMAKE_SYSROOT
for these tests. This patch implements it.

vitalybuka commented on compiler-rt/test/msan/lit.cfg.py:
```diff
@@ -17,7 +17,9 @@
     + [config.target_cflags]
-    + config.debug_info_flags
+    + ["--sysroot", config.cmake_sysroot]
+    if config.cmake_sysroot and config.host_os == "Linux"
```
should we do that for entire compiler-rt?
https://github.com/llvm/llvm-project/blob/main/compiler-rt/test/lit.common.configured.in

zeroomega replied:
I tried but ran into some further issues as some rtsan unit tests are not compatible with older glibc as `memfd_create` was not defined in glibc 2.19, it still can be invoked through syscall wrappers but that probably needs a separate patch. Therefore, I current limit the change to msan on Linux only to avoid other unwanted breakages.

vitalybuka replied:
It's seems like incremental which which does not lead us to desired direction.
Could you try to figure out rtsan issue instead?

petrhosek commented on compiler-rt/lib/msan/tests/CMakeLists.txt:
```diff
@@ -63,7 +63,11 @@ set(MSAN_UNITTEST_LINK_FLAGS
+  append_list_if(CMAKE_SYSROOT "--sysroot=${CMAKE_SYSROOT}" MSAN_UNITTEST_LINK_FLAGS)
+endif()
+
+  append_list_if(COMPILER_RT_HAS_LIBDL -ldl MSAN_UNITTEST_LINK_FLAGS)
```
```suggestion
append_list_if(COMPILER_RT_HAS_LIBDL -ldl MSAN_UNITTEST_LINK_FLAGS)
```

petrhosek commented on compiler-rt/lib/msan/tests/CMakeLists.txt:
```diff
@@ -63,7 +63,11 @@ set(MSAN_UNITTEST_LINK_FLAGS
 )
 
-append_list_if(COMPILER_RT_HAS_LIBDL -ldl MSAN_UNITTEST_LINK_FLAGS)
+if (LINUX)
```
Do we need to restrict this to `LINUX`?

petrhosek commented on compiler-rt/test/msan/lit.cfg.py:
```diff
@@ -18,6 +18,11 @@
     ]
     + [config.target_cflags]
     + config.debug_info_flags
+    + (
+        ["--sysroot", config.cmake_sysroot]
+        if config.cmake_sysroot and config.host_os == "Linux"
+        else []
+    )
```
I'd move this to a separate `if` block similar to other blocks below for readability.

petrhosek commented on compiler-rt/lib/msan/tests/CMakeLists.txt:
```diff
@@ -63,7 +63,11 @@ set(MSAN_UNITTEST_LINK_FLAGS
 
-append_list_if(COMPILER_RT_HAS_LIBDL -ldl MSAN_UNITTEST_LINK_FLAGS)
+if (LINUX)
+  append_list_if(CMAKE_SYSROOT "--sysroot=${CMAKE_SYSROOT}" MSAN_UNITTEST_LINK_FLAGS)
```
This needs to be included in `MSAN_UNITTEST_COMMON_CFLAGS` as well.

petrhosek commented on compiler-rt/test/msan/lit.site.cfg.py.in:
```diff
@@ -6,6 +6,7 @@ config.target_cflags = "@MSAN_TEST_TARGET_CFLAGS@"
 config.target_arch = "@MSAN_TEST_TARGET_ARCH@"
 config.use_lld = @MSAN_TEST_USE_LLD@
 config.use_thinlto = @MSAN_TEST_USE_THINLTO@
+config.cmake_sysroot = "@CMAKE_SYSROOT@"
```
Could we instead include the `--sysroot=${CMAKE_SYSROOT}` flag in `MSAN_TEST_TARGET_CFLAGS`?

zeroomega replied:
I tried but run into an issue with lit's glibc features detection and it caused segfault when running test: preinit_array.cpp

The reason is that `MSAN_TEST_TARGET_CFLAGS` is used when detecting the test's runtime glibc version: https://github.com/llvm/llvm-project/blob/118757af11954ffd43c16a41e39348e0907bba92/compiler-rt/test/lit.common.cfg.py#L682

Putting the --sysroot into the target cflag causes the 'unsupported' tag to fail in 
https://github.com/llvm/llvm-project/blob/118757af11954ffd43c16a41e39348e0907bba92/compiler-rt/test/msan/preinit_array.cpp#L4 

Because now the check only applies to the build time glibc version, not the runtime one.

zeroomega committed:
amend! [compiler-rt] Propagate sysroot from CMake to msan tests

[compiler-rt] Propagate sysroot from CMake to msan tests

Some msan tests requires rpc/xdr.h, which is no longer available
in some newer glibc. In build systems with hermetic sysroot set through
CMake, this issue can be solved by setting the sysroot to CMAKE_SYSROOT
for these tests. This patch implements it.

zeroomega committed:
amend! amend! [compiler-rt] Propagate sysroot from CMake to msan tests

[compiler-rt] Propagate sysroot from CMake to msan tests

Some msan tests requires rpc/xdr.h, which is no longer available
in some newer glibc. In build systems with hermetic sysroot set through
CMake, this issue can be solved by setting the sysroot to CMAKE_SYSROOT
for these tests. This patch implements it.

petrhosek commented on compiler-rt/test/msan/lit.site.cfg.py.in:
```diff
@@ -6,6 +6,7 @@ config.target_cflags = "@MSAN_TEST_TARGET_CFLAGS@"
 config.target_arch = "@MSAN_TEST_TARGET_ARCH@"
 config.use_lld = @MSAN_TEST_USE_LLD@
 config.use_thinlto = @MSAN_TEST_USE_THINLTO@
+config.cmake_sysroot = "@CMAKE_SYSROOT@"
```
This is just a nit, but I'd suggest calling it just `sysroot`, the fact that it comes from `CMake` isn't relevant.
```suggestion
config.sysroot = "@CMAKE_SYSROOT@"
```

vitalybuka reviewed (approved):
if it's good enough @petrhosek, then it's good enough for me. LGTM
